### ISSUE TITLE
[jvm-package][python-package] inconsistencies in different api

### ISSUE THREAD

**Issue by geoHeil** [dmlc/xgboost#1992]
Using xgboost in java (xgboost4j) on spark and python seem to produce different results for me https://github.com/geoHeil/xgboostPythonVsSpark how can this be true? Is there something wrong with my setup?

## Environment info
Operating System: 0sx 10.12.2

Compiler: gcc-6

Package used (python/jvm):  both,
Python version is 3.5.2 :: Anaconda custom (x86_64)

`xgboost` version used: 5d74578095e1414cfcb62f9732165842f25b81ca

## Steps to reproduce

1. https://github.com/geoHeil/xgboostPythonVsSpark

**Comment by geoHeil** [comment]
@xydrolase  can you reproduce this strange behaviour? Do you have a hint what could be wrong?

**Comment by xydrolase** [comment]
I haven't replicated your experiments, but the code in your Jupyter notebook is somewhat worrying at first glance.

```python
    X_train['target'] = y_train
    X_test['target'] = y_test
```

I assume that you put the labels/targets into `X_train` and `X_test` for the purpose of dumping these matrices into CSV files. 

And then, 

```python
clf.fit(X_train, y_train,
                      eval_set=fit_params['eval_set'], eval_metric=fit_params['eval_metric'],
                      early_stopping_rounds=fit_params['early_stopping_rounds'], verbose=False)
```

Because `X_train` contains `y_train`, surely it will overfit. And I believe if you print out the feature importance, `target` will be the most important feature, because it always predict `y_train` perfectly. 

I didn't look at your Scala code, but I think the above is enough to explain the overfit issue.

**Comment by geoHeil** [comment]
@xydrolase sorry for that. this is a problem with the example. Indeed this finally shows similar results for the minimal example.

Unfortunately, this will not help me to find the problem in my **real** code where I am facing a similar issue where the python variant will perform a) considerably better and b) consistently better. 

Will need to close this issue, but am looking forward to any suggestions what might be a problem causing this issue in my real code.

**Comment by xydrolase** [comment]
When you ran your code in Spark, did you run it in a distributed fashion, or just using one executor?
I guess to pinpoint the root cause of your issue, have you tried to use Scala XGBoost directly to train a model?

**Comment by geoHeil** [comment]
Was just on a single executor. So far only the xgboost4j variant. Do you
think the scala version would be different?

Xin Yin <notifications@github.com> schrieb am Mo., 30. Jan. 2017 um
13:29 Uhr:

**Comment by geoHeil** [comment]
Was just on a single executor. So far only the xgboost4j variant. Do you think the scala version would be different?

**Comment by xydrolase** [comment]
No, I was just trying to figure out what might be the cause of the inconsistency that you've observed (be it JVM, or Spark.) Since there are `ml.dmlc.xgboost4j.scala.XGBoost` and `ml.dmlc.xgboost4j.scala.spark.XGBoost`, I wonder if you can replicate the same issue using the former. 

Also, how did you load the data into Spark?

**Comment by xydrolase** [comment]
Also, IIRC, the JVM version of XGBoost does not support early stopping. If you've noticed that your Python variant stopped fairly early, then it could be that the JVM variant over fitted due to your  model parameters. 

Since you got identical results on the simulated dataset, perhaps the issue is not with the implementation itself, but subtle differences in how did you use the two different implementations.

**Comment by geoHeil** [comment]
Initial data is a CSV. Then cleaning, preprocessing, null handling and some feature engineering is performed.

The whole pipeline looks like as outlined below for the fit stage:
1. preprocessor (null handling, some basic feature engineering)
2. categorical handler (fit one string indexer per categorical column, additionally one-hot encode, learn most frequent labels per column)
3. calculate (effect coding) %-tage values to handle categorical variables with too many levels to be dummy coded
4. vector assembler
5. xgboost

For the transform step
1. preprocessor (just the same)
2. categorical handler: will figure out any new labels and replace with most frequent, then go on as before
3, 4,5, same as before but in transform mode.

The interesting thing is:
for some folds, xgboost4j is getting a similar score (kappa) than the python variant. For most folds, it will be a lot worse. When I tried to generate the folds in spark and load these into python the results were (not as good as the plain python variant, obviously different split) but consistently at an acceptable score (kappa), and not jumping a lot between the folds. 
 
Indeed you are correct my real code uses early stopping. But I can say that the xgbost4j code is far from overfitting the data.

Besides early stopping do you see any other subtle differences?

**Comment by xydrolase** [comment]
Do you have separate implementations in Python and Spark for the data preprocessing? 
By folds, did you mean you used K-fold CV? So did you use Spark and sklearn accordingly when creating these CV folds? The data shuffling can certainly be different due the randomness, but you did mention that you noticed better results when you loaded the Spark data into Python and trained it with Python.

Still, there are too many variables to reach to a conclusive argument where exactly it might go wrong. If you indeed have two separate preprocessing pipelines, I suggest you go with a plain training/validation split. Finish all your preprocessing in Python, save the dataset into CSV files, load them into Spark, train a model, and validate the performance on the validation set. And turn off early-stopping Python. In this way, you can isolate the cause of the issue.

**Comment by geoHeil** [comment]
@xydrolase I will definitely try turning off early stopping. Ans will notify you here about my results.

**Comment by geoHeil** [comment]
@xydrolase it is correct to assume the same settings (i.e. seed) same splits (all from python as CSV) should produce the equal numbers? For one (little bit bigger) example they differ on the third digit.

**Comment by xydrolase** [comment]
Seed for Python package, or seed for XGBoost?
You should have the same splits, if you provide the same split. 
Gradient boost tree has its own randomization (sampling of columns/rows). Without these sources of stochasticity, the tree-splitting process should be deterministic. I haven't tested setting the same seed in XGBoost for reproducibility, but if in various runs you only noticed differences on the 3rd digit, then I wouldn't expect the Scala's results to have a much larger variance.

**Comment by geoHeil** [comment]
@xydrolase when trying to reduce the pipeline to the minimum (not handling categorical values, only numeric data) I noticed that there are minimal deviations between python and spark.
A split in python e.g. will give me 50.9 whereas the very same split in spark will give me 53.2 in xgboost4j.

I attribte this minor difference to the different implementations of random number generators? Is this sensible? Shouldn't the same C binary of xgboost get the same seed and perform the same calculations?

**Comment by geoHeil** [comment]
To clarify seed for both is set to the same value.
